### PR TITLE
Fix Issue 18766 - std.typecons.Tuple.toHash could be improved

### PR DESCRIPTION
**Pull request by n8sh**
Current std.typecons.Tuple has a custom `toHash` that is just the sum of the hashes of the tuple's fields. This PR instead combines values using ~~the runtime's default hash function~~ the equivalent of [`boost::hash_combine`](https://www.boost.org/doc/libs/1_55_0/doc/html/hash/reference.html#boost.hash_combine).

### CODE REVIEW

**Comment by n8sh** [comment]
On the other hand it might be even better to delete the custom `toHash`.

**Comment by JackStouffer** [comment]
If it has a custom opEquals it needs a toHash

**Comment by n8sh** [comment]
I see.

**Comment by n8sh** [comment]
Updated to use the equivalent of [`boost::hash_combine`](https://www.boost.org/doc/libs/1_55_0/doc/html/hash/reference.html#boost.hash_combine).

File: std/typecons.d [hunk 1]
```diff
@@ -1177,8 +1177,16 @@ if (distinctFieldNames!(Specs))
         size_t toHash() const nothrow @trusted
         {
             size_t h = 0;
-            foreach (i, T; Types)
-                h += typeid(T).getHash(cast(const void*)&field[i]);
+            static foreach (i, T; Types)
+            {{
+                const k = typeid(T).getHash(cast(const void*) &field[i]);
```
Comment (JackStouffer) [inline std/typecons.d]: Should use trusted escape here instead of on the whole function

Comment (n8sh) [reply std/typecons.d]: done

File: std/typecons.d [hunk 2]
```diff
@@ -1174,11 +1174,19 @@ if (distinctFieldNames!(Specs))
...
          */
-        size_t toHash() const nothrow @trusted
+        size_t toHash() const nothrow @safe
         {
             size_t h = 0;
-            foreach (i, T; Types)
-                h += typeid(T).getHash(cast(const void*)&field[i]);
+            static foreach (i, T; Types)
+            {{
+                const k = typeid(T).getHash((() @trusted => cast(const void*) &field[i])());
```
Comment (edi33416) [inline std/typecons.d]: Shouldn't this call `T.toHash` if it's defined?

File: std/typecons.d [hunk 3]
```diff
@@ -1174,11 +1174,19 @@ if (distinctFieldNames!(Specs))
...
+                static if (i == 0)
+                    h = k;
+                else
+                    // As in boost::hash_combine
+                    // https://www.boost.org/doc/libs/1_55_0/doc/html/hash/reference.html#boost.hash_combine
+                    h ^= k + 0x9e3779b9 + (h << 6) + (h >>> 2);
```
Comment (edi33416) [inline std/typecons.d]: Maybe we could use `boost::hash_combine` to enhance how the hash of an `AA` is computed by [`hashOf`](https://github.com/dlang/druntime/blob/master/src/core/internal/hash.d#L170)?
Currently that just does a `^=`

File: std/typecons.d [hunk 4]
```diff
@@ -1174,11 +1174,19 @@ if (distinctFieldNames!(Specs))
             Returns:
                 A `size_t` representing the hash of this `Tuple`.
          */
-        size_t toHash() const nothrow @trusted
+        size_t toHash() const nothrow @safe
```
Comment (edi33416) [inline std/typecons.d]: Can this also be `pure`? I believe it should be.

File: std/typecons.d (hunk 2 above)
Comment (n8sh) [reply std/typecons.d]: Yes, it should:

```d
class TypeInfo_Struct : TypeInfo
{
//...
    override size_t getHash(scope const void* p) @trusted pure nothrow const
    {
        assert(p);
        if (xtoHash)
        {
            return (*xtoHash)(p);
        }
        else
        {
            import core.internal.traits : externDFunc;
            alias hashOf = externDFunc!("rt.util.hash.hashOf",
                                        size_t function(const(void)[], size_t) @trusted pure nothrow @nogc);
            return hashOf(p[0 .. initializer().length], 0);
        }
    }
//...
}
```

File: std/typecons.d (hunk 4 above)
Comment (n8sh) [reply std/typecons.d]: It can't because the base signature of `TypeInfo.getHash` doesn't include `pure`.

File: std/typecons.d (hunk 3 above)
Comment (n8sh) [reply std/typecons.d]: We can't in that case because the AA's hash needs to be independent of traversal order.

Comment (wilzbach) [reply std/typecons.d]: Out of interest: why not use `std.digest` (e.g. CRC) and use the nicer `.put` API?

**Pull request merged**